### ISSUE TITLE
Simplify our promote code.

### ISSUE THREAD

**Issue by vext01** [ykjit/yk#777]
I was investigating why `yk_promote(x)` (instead of `x = yk_promote(x)`) still promotes `x`.

In our initial design we intended for only the returned value to be promoted and uses of the original `x` should remain unpromoted. Turns out that's not gonna work.

The reason is, when you do `yk_promote(x)`, regardless of whether you capture the returned value or not you get a value guard, and the guard itself gives the LLVM optimiser the ability to assume the promoted value of `x` already.

If we have:

```
define void @f(i32 %0) {
  yk_promote(%0)
  %1 = add i32 %0,  1
} 
```

And you trace this when `%0` is `5`, then you get a **pre-opt** trace like:
```
%1 = icmp ne i32 %0, 5
br i1 %1, %guardfail, %guardsuccess
...

guardsuccess:
  %2 = add i32 %0, 5
```

If you give this to LLVM it figures out that in the `guardsuccess` block `%0=5` and the **post-opt** trace then looks like:
```
%1 = icmp ne i32 %0, 5
br i1 %1, %guardfail, %guardsuccess
...

guardsuccess:
  %2 = i32 6
```

In other words, with sufficient optimisations enabled (this only started happening in my experimental AOT optimisations branch), LLVM will "promote" the argument to `yk_promote` anyway. In general we can't stop it.

What does this mean:
  - `x = yk_promote(y)` can promote `y`, so what's the point of returning `x`?
  -  We can simplify the trace compiler.

About the latter:

Upon encountering `y = yk_promote(x)` the trace compiler (currently):
 - emits a value guard for `x`
 - rewrites all future uses of `y` to be a constant.

As we've shown, with sufficient optimisations, no rewriting is necessary.
